File: autocontext/src/autocontext/integrations/browser/context_capture.py
"""Reusable browser snapshot capture helpers."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from pathlib import Path

from autocontext.config.settings import AppSettings
from autocontext.integrations.browser.contract.models import BrowserSessionConfig
from autocontext.integrations.browser.factory import browser_runtime_from_settings
from autocontext.integrations.browser.types import BrowserRuntimePort

MAX_BROWSER_VISIBLE_TEXT_CHARS = 1200


@dataclass(frozen=True, slots=True)
class CapturedBrowserContext:
    """Stable browser-derived context that can be folded into prompts."""

    url: str
    title: str
    visible_text: str
    html_path: str | None = None
    screenshot_path: str | None = None


def capture_browser_context(
    settings: AppSettings,
    *,
    browser_url: str,
    evidence_root: Path,
) -> CapturedBrowserContext:
    """Capture a single browser snapshot and normalize its text payload."""
    configured = browser_runtime_from_settings(settings, evidence_root=evidence_root)
    if configured is None:
        raise ValueError("browser exploration is disabled")

    return asyncio.run(
        _capture_browser_context_async(
            configured.runtime,
            configured.session_config,
            browser_url=browser_url,
        )
    )


async def _capture_browser_context_async(
    runtime: BrowserRuntimePort,
    session_config: BrowserSessionConfig,
    *,
    browser_url: str,
) -> CapturedBrowserContext:
    session = await runtime.create_session(session_config)
    try:
        navigation = await session.navigate(browser_url)
        if not navigation.allowed:
            raise ValueError(f"browser navigation blocked by policy: {navigation.policyReason}")
        snapshot = await session.snapshot()
    finally:
        await session.close()

    return CapturedBrowserContext(
        url=snapshot.url,
        title=snapshot.title,
        visible_text=_trim_visible_text(snapshot.visibleText),
        html_path=snapshot.htmlPath,
        screenshot_path=snapshot.screenshotPath,
    )


def render_captured_browser_context(context: CapturedBrowserContext) -> str:
    """Render browser context into prompt-friendly lines."""
    lines = [
        "Live browser context:",
        f"URL: {context.url}",
        f"Title: {context.title}",
        f"Visible text: {context.visible_text}",
    ]
    if context.html_path:
        lines.append(f"HTML artifact: {context.html_path}")
    if context.screenshot_path:
        lines.append(f"Screenshot artifact: {context.screenshot_path}")
    return "\n".join(lines)


def _trim_visible_text(text: str) -> str:
    normalized = " ".join(text.split())
    if len(normalized) <= MAX_BROWSER_VISIBLE_TEXT_CHARS:
        return normalized
    return normalized[:MAX_BROWSER_VISIBLE_TEXT_CHARS].rstrip()


__all__ = [
    "CapturedBrowserContext",
    "capture_browser_context",
    "render_captured_browser_context",
]

File: ts/src/integrations/browser/context-capture.ts
import { createBrowserRuntimeFromSettings, type BrowserRuntimeSettingsLike } from "./factory.js";

const MAX_BROWSER_VISIBLE_TEXT_CHARS = 1200;

export interface CapturedBrowserContext {
  readonly url: string;
  readonly title: string;
  readonly visibleText: string;
  readonly htmlPath?: string | null;
  readonly screenshotPath?: string | null;
}

export type BrowserContextCaptureSettingsLike = BrowserRuntimeSettingsLike;

export interface CaptureBrowserContextRequest {
  readonly settings: BrowserContextCaptureSettingsLike;
  readonly browserUrl: string;
  readonly evidenceRoot: string;
}

export interface BrowserContextCaptureDependencies {
  readonly createBrowserRuntimeFromSettings: typeof createBrowserRuntimeFromSettings;
}

const DEFAULT_DEPENDENCIES: BrowserContextCaptureDependencies = {
  createBrowserRuntimeFromSettings,
};

export async function captureBrowserContextFromUrl(
  opts: CaptureBrowserContextRequest,
  dependencies: BrowserContextCaptureDependencies = DEFAULT_DEPENDENCIES,
): Promise<CapturedBrowserContext> {
  const configured = dependencies.createBrowserRuntimeFromSettings(opts.settings, {
    evidenceRoot: opts.evidenceRoot,
  });
  if (!configured) {
    throw new Error("browser exploration is disabled");
  }

  const session = await configured.runtime.createSession(configured.sessionConfig);
  try {
    const navigation = await session.navigate(opts.browserUrl);
    if (!navigation.allowed) {
      throw new Error(`browser navigation blocked by policy: ${navigation.policyReason}`);
    }

    const snapshot = await session.snapshot();
    return {
      url: snapshot.url,
      title: snapshot.title,
      visibleText: trimCapturedBrowserText(snapshot.visibleText),
      htmlPath: snapshot.htmlPath ?? null,
      screenshotPath: snapshot.screenshotPath ?? null,
    };
  } finally {
    await session.close();
  }
}

export function renderCapturedBrowserContext(context: CapturedBrowserContext): string {
  const lines = [
    "Live browser context:",
    `URL: ${context.url}`,
    `Title: ${context.title}`,
    `Visible text: ${context.visibleText}`,
  ];
  if (context.htmlPath) {
    lines.push(`HTML artifact: ${context.htmlPath}`);
  }
  if (context.screenshotPath) {
    lines.push(`Screenshot artifact: ${context.screenshotPath}`);
  }
  return lines.join("\n");
}

function trimCapturedBrowserText(text: string): string {
  const normalized = text.replace(/\s+/g, " ").trim();
  return normalized.length <= MAX_BROWSER_VISIBLE_TEXT_CHARS
    ? normalized
    : normalized.slice(0, MAX_BROWSER_VISIBLE_TEXT_CHARS).trimEnd();
}
